### PR TITLE
Add Ubernetes Lite e2e tests for spreading RC and Service pods evenly…

### LINKED ISSUE #19415: Add Ubernetes Lite e2e test for spreading RC pods spreading evenly across all zones
Part of #19413

### PR DESCRIPTION
**Pull request by ghost**
… across all zones.

Fixes #19415

I've run this a few hundred times against a private test cluster and it passed consistently.

### CODE REVIEW

**Comment by quinton-hoole** [comment]
@justinsb FYI

**Comment by ihmccreery** [comment]
@quinton-hoole I'm happy to review this, but why me?  This seems like something someone from the control-plane team should see.  I have ~no context on Ubernetes Lite.

**Comment by justinsb** [comment]
I can review if you'd like - assign to me if so!

File: test/e2e/util.go [hunk 1]
```diff
@@ -1791,6 +1801,28 @@ func (config *RCConfig) start() error {
 	return nil
 }
 
+// Simplified version of RunRC, that does not create RC, but creates plain Pods.
+// optionally waits for pods to start running (if waitForRunning == true)
+func startPods(c *client.Client, replicas int, namespace string, podNamePrefix string, pod api.Pod, waitForRunning bool) {
+	// var podNames []string
```
Comment (justinsb) [inline test/e2e/util.go]: Probably remove the podNames comments out code

File: test/e2e/util.go [hunk 2]
```diff
@@ -1791,6 +1801,28 @@ func (config *RCConfig) start() error {
...
+		// podNames := append(podNames, podName)
+		pod.ObjectMeta.Name = podName
+		pod.ObjectMeta.Labels["name"] = podName
+		pod.ObjectMeta.Labels["startPodsID"] = startPodsID
+		pod.Spec.Containers[0].Name = podName
+		_, err := c.Pods(namespace).Create(&pod)
+		expectNoError(err)
+	}
+	if waitForRunning {
+		label := labels.SelectorFromSet(labels.Set(map[string]string{"startPodsID": startPodsID}))
```
Comment (justinsb) [inline test/e2e/util.go]: This seems a huge improvement for parallel tests over what we had before...

File: test/e2e/util.go [hunk 3]
```diff
@@ -1948,6 +1980,41 @@ waitLoop:
...
+	}
+	return true, nil
+}
+
+// Wait for all matching pods to become scheduled and at least one
+// matching pod exists.  Return the list of matching pods.
+func waitForPodsWithLabelScheduled(c *client.Client, ns string, label labels.Selector) (pods *api.PodList, err error) {
+	err = wait.PollImmediate(poll, podScheduledBeforeTimeout,
+		func() (bool, error) {
+			pods, err = waitForPodsWithLabel(c, ns, label)
```
Comment (justinsb) [inline test/e2e/util.go]: I'm a little confused about the timeout here .... so we poll for 20 seconds max, but the inner poll function has a 10 minute timeout?

File: test/e2e/ubernetes_lite.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/intstr"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+var _ = Describe("Ubernetes Lite", func() {
```
Comment (ihmccreery) [inline test/e2e/ubernetes_lite.go]: Presumably you want this to be [Feature:UbernetesLite] or something similar?  See https://github.com/kubernetes/kubernetes/blob/master/docs/devel/e2e-tests.md#kinds-of-tests.

**Comment by ihmccreery** [comment]
Thanks @justinsb.  Only one comment from me, about labeling this test appropriately.

**Comment by justinsb** [comment]
A few nits @quinton-hoole but otherwise LGTM.  Will you have time to address them, or should we merge and you can address the nits later?

File: test/e2e/ubernetes_lite.go (hunk 4 above)
Comment (quinton-hoole) [inline test/e2e/ubernetes_lite.go]: No, this is actually core functionality, and one of the main features of our upcoming 1.2 release. 
The code on line 45 causes the tests to be automatically skipped on non-multi-zone clusters (which all our test clusters except one - "ubernetes-lite" - are, currently).
As soon as all our e2e tests have proved to be stable against multi-zone clusters (they are all currently passing as we speak), I intend to add the "ubernetes-lite" multi-zone test cluster to the list of merge-queue blocking test clusters.

File: test/e2e/util.go (hunk 1 above)
Comment (quinton-hoole) [inline test/e2e/util.go]: Yup, done.

File: test/e2e/util.go (hunk 2 above)
Comment (quinton-hoole) [inline test/e2e/util.go]: Why thank you :-)

File: test/e2e/util.go (hunk 3 above)
Comment (quinton-hoole) [inline test/e2e/util.go]: Agreed - that's messed up.  Let me fix that now, and then I think we're ready for LGTM.

**Comment by quinton-hoole** [comment]
@justinsb I've addressed all your nits, I think. PTAL.

File: test/e2e/util.go (hunk 3 above)
Comment (quinton-hoole) [inline test/e2e/util.go]: The 10 minute timeout comment was rubbish - I fixed that.  The actual value is currently 1 minute. And I made podScheduledBeforeTimeout be podListTimout plus a scheduling delay.

**Comment by justinsb** [comment]
Awesome - thanks!  LGTM

**Comment by quinton-hoole** [comment]
@fabioy FYI I'm not sure why the e2e 'errored' message occurs above, as according to Jenkins that run passed just fine. Perhaps we can merge this one manually?

```

Ran 137 of 239 Specs in 483.695 seconds
SUCCESS! -- 137 Passed | 0 Failed | 0 Pending | 102 Skipped 
...
*** View logs and artifacts at https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/pull/21474/kubernetes-pull-build-test-e2e-gce/30504 ***


[Google Cloud Storage Uploader] Uploading: build-log.txt
Setting status of a9fd20714036dc754cd6ab443f2523fc36b0e35d to SUCCESS with url http://pr-test.k8s.io/21474/kubernetes-pull-build-test-e2e-gce/30504/ and message: ' 239 tests run, 102 skipped, 0 failed.'
Using context: Jenkins GCE e2e
GCE e2e build/test **errored** for commit a9fd20714036dc754cd6ab443f2523fc36b0e35d.
* [Build Log](http://pr-test.k8s.io/21474/kubernetes-pull-build-test-e2e-gce/30504/build-log.txt)
* [Test Artifacts](https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/21474/kubernetes-pull-build-test-e2e-gce/30504/artifacts/)
* [Internal Jenkins Results](http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//30504)

Please reference the [list of currently known flakes](https://github.com/kubernetes/kubernetes/issues?q=is:issue+label:kind/flake+is:open) when examining this failure.

[WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
Finished: SUCCESS


```

**Comment by ixdy** [comment]
@quinton-hoole I think I fat-fingered the Jenkins configuration when adding that flake link. Fixed now.

**Comment by quinton-hoole** [comment]
Urgh, looks like a transient gcloud error.  I don't think that we do proper retries in any of our cloud up scripts.

Manually retrying.

@k8s-bot test this 

```
Created [https://www.googleapis.com/compute/v1/projects/kubernetes-jenkins-pull/global/firewalls/e2e-gce-builder-2-2-master-https].
NAME                             NETWORK             SRC_RANGES RULES   SRC_TAGS TARGET_TAGS
e2e-gce-builder-2-2-master-https e2e-gce-builder-2-2 0.0.0.0/0  tcp:443          e2e-gce-builder-2-2-master
ERROR: (gcloud.compute.addresses.describe) Could not fetch resource:
 - Backend Error
 
2016/02/25 14:50:48 e2e.go:278: Error running up: exit status 1
2016/02/25 14:50:48 e2e.go:274: Step 'up' finished in 35.908786012s
2016/02/25 14:50:48 e2e.go:140: Error starting e2e cluster. Aborting.
exit status 1
```

**Comment by fabioy** [comment]
@quinton-hoole Checks passed. Do you still want me to manually merge?

**Comment by quinton-hoole** [comment]
Yup, that would be great.

**Pull request merged**